peekingduck/pipeline/utils/bbox/transforms.py

```diff
@@ -0,0 +1,96 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility functions which convert bbox coordinates from one format to
+another.
+"""
+
+import numpy as np
+import torch
+
+
+def tlwh2xyah(inputs: np.ndarray) -> np.ndarray:
+    """Converts bounding box to format `(center x, center y, aspect ratio,
+    height)`, where the aspect ratio is `width / height`.
+
+    Args:
+        tlwh (np.ndarray): Input bounding box with format `(top left x,
+            top left y, width, height)`.
+    Returns:
+        (np.ndarray): Bounding box with (x, y, a, h) format.
+    """
+    outputs = np.asarray(inputs).copy()
+    outputs[:2] += outputs[2:] / 2
+    outputs[2] /= outputs[3]
+    return outputs
+
+
+def tlwh2xyxyn(inputs: np.ndarray, height: int, width: int) -> np.ndarray:
+    """Converts from [t, l, w, h] to [x1, y1, x2, y2] format.
+
+    (x1, y1) and (x2, y2) are coordinates of top left and bottom right
+    respectively. (t, l) is the coordinates of the top left corner, w is the
+    width, and h is the height.
+    """
+    outputs = np.empty_like(inputs)
+    outputs[:, 0] = inputs[:, 0] / width
+    outputs[:, 1] = inputs[:, 1] / height
+    outputs[:, 2] = (inputs[:, 0] + inputs[:, 2]) / width
+    outputs[:, 3] = (inputs[:, 1] + inputs[:, 3]) / height
+    return outputs
+
+
+def xywh2xyxy(inputs: torch.Tensor) -> torch.Tensor:
+    """Converts from [x, y, w, h] to [x1, y1, x2, y2] format.
+
+    (x, y) is the object center. (x1, y1) is the top left corner and (x2, y2)
+    is the bottom right corner.
+    """
+    outputs = torch.empty_like(inputs)
+    outputs[:, 0] = inputs[:, 0] - inputs[:, 2] / 2
+    outputs[:, 1] = inputs[:, 1] - inputs[:, 3] / 2
+    outputs[:, 2] = inputs[:, 0] + inputs[:, 2] / 2
+    outputs[:, 3] = inputs[:, 1] + inputs[:, 3] / 2
+
+    return outputs
+
+
+def xyxy2tlwh(inputs: np.ndarray) -> np.ndarray:
+    """Converts bounding box to format `(top left x, top left y, width,
+    height)`.
+
+    Args:
+        inputs (np.ndarray): Input bounding box with format (x1, y1, x2, y2)
+            where (x1, y1) is top left, (x2, y2) is bottom right.
+
+    Returns:
+        (np.ndarray): Bounding box with `(top left x, top left y, width,
+            height)` format.
+    """
+    outputs = np.asarray(inputs).copy()
+    outputs[2:] -= outputs[:2]
+    return outputs
+
+
+def xyxy2xyxyn(inputs: np.ndarray, height: float, width: float) -> np.ndarray:
+    """Converts from [x1, y1, x2, y2] to normalized [x1, y1, x2, y2].
+
+    (x1, y1) is the top left corner and (x2, y2) is the bottom right corner.
+    Normalized coordinates are w.r.t. original image size.
+    """
+    outputs = np.empty_like(inputs)
+    outputs[:, [0, 2]] = inputs[:, [0, 2]] / width
+    outputs[:, [1, 3]] = inputs[:, [1, 3]] / height
+
+    return outputs
```

Comment:
For all the above functions:
1. Would be instructive to users if docstrings could include a numeric example showing the input and output.
2. For each x, y, w, h, t, l : if there are expectations for any to be in the range (0, 1) (relative vs abs coords), would be good to have assert statements to check for input range errors.
3. Some functions (e.g. `tlwh2xyah`) accept single dimension input, e.g. [0, 0, 1, 1], while others (e.g. `tlwh2xyxyn`) require multi-dimension input, e.g. [[0, 0, 1, 1]]. Would be good to document this somewhere in the docstrings.

Reply:
1. Added example input/output for each function
2. For now, these functions have not expectation for input to be of a specific range. Will use assert statements if necessary in the future
3. Specified in the `Args` section of the docstring that `inputs` is either a 1-d array or 2-d array.